### PR TITLE
Update minimum python to 3.8

### PR DESCRIPTION
**Pull request by dbczumar**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Update minimum python to 3.8

## How is this patch tested?

Existing CI

## Release Notes

Update minimum python to 3.8

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [X] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [X] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by harupy** [comment]
We're going to merge this PR to branch-2.0, then merge branch 2.0?

**Comment by dbczumar** [comment]
> We're going to merge this PR to branch-2.0, and then merge branch 2.0 to master via Ben's PR?

I'll probably wait till @BenWilson2 's PR is merged and then retarget this PR against `master`.

**Comment by harupy** [comment]
Makes sense :)

**Comment by harupy** [comment]
@dbczumar Cancelled the cross version tests since they were hanging.

**Review by BenWilson2 — APPROVED** [review]
LGTM!

File: .github/actions/setup-python/action.yml [hunk 1]
```diff
@@ -23,16 +23,14 @@ runs:
...
           if [ ${{ runner.os }} == "Windows" ]; then
-            python_version="3.7.9"
+            # GitHub Actions does not have a Python 3.8.12 binary for Windows as of Oct. 27 2022
+            python_version="3.8.10"
           else
-            python_version="3.7.12"
-          fi
-        elif [[ "$python_version" == "3.8" ]]; then
-          python_version="3.8.12"
+            python_version="3.8.12"
```
Comment (BenWilson2) [inline .github/actions/setup-python/action.yml:31]: (add in `fi` to close the if/else/fi loop)

Comment (dbczumar) [reply .github/actions/setup-python/action.yml:31]: Thanks @BenWilson2 !

**Comment by BenWilson2** [comment]
Should we revisit the R Keras test and move it from the Keras API to the TF API in R? That TF pinning to support that R test suite isn't compatible with py3.8, right?

**Comment by harupy** [comment]
[scikit-learn 1.1.0](https://github.com/scikit-learn/scikit-learn/releases/tag/1.1.0) caused some tests to fail.

**Comment by harupy** [comment]
https://github.com/mlflow/mlflow/actions/runs/3340751160/jobs/5531207012#step:8:26

```
Error: tests/conftest.py:148:0: W0240: Class 'ExtendedMonkeyPatch' is a subclass of a class decorated with typing.final: 'MonkeyPatch' (subclassed-final-class)
```

https://pylint.pycqa.org/en/latest/user_guide/messages/warning/subclassed-final-class.html

This rule is only enabled in Python >= 3.8. Let's ignore this rule.

**Comment by harupy** [comment]
I'll merge https://github.com/mlflow/mlflow/pull/7142 once all the CI checks pass. Is this PR a final blocker for MLflow 2.0 RC?

**Comment by dbczumar** [comment]
> I'll merge #7142 once all the CI checks pass. Is this PR a final blocker for MLflow 2.0 RC?

Awesome! Thanks @harupy ! Yes, this PR and #7142 are the final blockers for MLflow 2.0 RC unless you can think of anything else :)

**Comment by dbczumar** [comment]
> > I'll merge #7142 once all the CI checks pass. Is this PR a final blocker for MLflow 2.0 RC?
> 
> Awesome! Thanks @harupy ! Yes, this PR and #7142 are the final blockers for MLflow 2.0 RC unless you can think of anything else :)

@sunishsheth2009 is also working on a couple of small runs table fixes (bringing back the models and run source columns). Are those going to be ready today?

**Comment by dbczumar** [comment]
> > > I'll merge #7142 once all the CI checks pass. Is this PR a final blocker for MLflow 2.0 RC?
> > 
> > 
> > Awesome! Thanks @harupy ! Yes, this PR and #7142 are the final blockers for MLflow 2.0 RC unless you can think of anything else :)
> 
> @sunishsheth2009 is also working on a couple of small runs table fixes (bringing back the models and run source columns). Are those going to be ready today / Monday?

@harupy I talked with @sunishsheth2009 offline. We'll have the small UI fixes merged by Monday and can release the RC as soon as that's done :)

**Comment by harupy** [comment]
@dbczumar @sunishsheth2009 Got it :)

**Comment by ramsey-superpedestrian** [comment]
I don't think this was a good idea.

Python 3.7 is still a supported version of python.  It's still receiving patches.  And there are lots of platforms that are locked to 3.7.  

The most notable one being Google Colab.

I'm going to open an installation issue.  Perhaps you would consider reintroducing python 3.7 support.  Otherwise, to use Google Colab, either users will have to downgrade to MLFlow 1.3.0 or we will have to fork and revert this PR missing all downstream changes.

**Comment by ramsey-superpedestrian** [comment]
I don't really understand why this constraint was added.  It means that users of platforms that are locked on 3.7 (i.e. Google Colab) cannot use the mlflow 2.0.1 client.

**Pull request merged**